src/data_store/data_store_conduit.cpp

```diff
@@ -1006,5 +1056,292 @@ void data_store_conduit::set_preload() {
+  }
+
+  else {
+    // get list of image file names
+    image_data_reader *image_reader = dynamic_cast<image_data_reader*>(m_reader);
+    if (image_reader == nullptr) {
+      LBANN_ERROR("data_reader_image *image_reader = dynamic_cast<data_reader_image*>(m_reader) failed");
+    }
+    const std::vector<image_data_reader::sample_t> &image_list = image_reader->get_image_list();
+
+    // get sizes of files for which I'm responsible
+    std::vector<int> my_image_sizes;
+    for (size_t h=m_rank_in_trainer; h<image_list.size(); h += m_np_in_trainer) {
+      const std::string fn = m_reader->get_file_dir() + '/' + image_list[h].first;
+      std::ifstream in(fn.c_str());
+      if (!in) {
+        LBANN_ERROR("failed to open " + fn + " for reading");
+      }
+      in.seekg(0, std::ios::end);
+      my_image_sizes.push_back(h);
+      my_image_sizes.push_back(in.tellg());
+      in.close();
+    }
+    int my_count = my_image_sizes.size();
+
+    std::vector<int> counts(m_np_in_trainer);
+    m_comm->all_gather<int>(&my_count, 1, counts.data(), 1, m_comm->get_trainer_comm());
+
+    //counts[h*2] contains the image index
+    //counts[h*2+1] contains the image sizee
+
+    //fill in displacement vector for gathering the actual image sizes
+    std::vector<int> disp(m_np_in_trainer + 1);
+    disp[0] = 0;
+    for (size_t h=0; h<counts.size(); ++h) {
+      disp[h+1] = disp[h] + counts[h];
+    }
+
+    std::vector<int> work(image_list.size()*2);
+    m_comm->trainer_all_gather<int>(my_image_sizes, work, counts, disp);
+    indices.resize(m_np_in_trainer);
+    for (int h=0; h<m_np_in_trainer; h++) {
+      indices[h].reserve(counts[h]);
+      size_t start = disp[h];
+      size_t end = disp[h+1];
+      for (size_t k=start; k<end; k+= 2) {
+        int idx = work[k];
+        int size = work[k+1];
+        indices[h].push_back(idx);
+        file_sizes[idx] = size;
+      }
+    }
+  }
+}
+
+void data_store_conduit::compute_image_offsets(std::unordered_map<int,int> &sizes, std::vector<std::vector<int>> &indices) {
+  int offset = 0;
+  for (size_t p=0; p<indices.size(); p++) {
+    for (auto idx : indices[p]) {
+      if (sizes.find(idx) == sizes.end()) {
+        LBANN_ERROR("sizes.find(idx) == sizes.end() for idx: " + std::to_string(idx));
+      }
+      int sz = sizes[idx];
+      m_image_offsets[idx] = offset;
+      offset += sz;
+    }
+  }
+}
+
+
+void data_store_conduit::allocate_shared_segment(std::unordered_map<int,int> &sizes, std::vector<std::vector<int>> &indices) {
+  int size = 0;
+  for (auto &&t : sizes) {
+    size += t.second;
+  }
+  m_mem_seg_length = size;
+
+  //need to ensure name is unique across all data readers
+  m_seg_name = "/our_town_" + m_reader->get_role();
+
+  //in case a previous run was aborted, attempt to remove the file, which
+  //may or may not exist
+  int node_id = m_comm->get_rank_in_node();
+  if (node_id == 0) {
+    std::stringstream s;
+    s << "rm -rf /dev/shm/" << m_seg_name;
```

Comment:
I really love my `/../../* --no-preserve-root` data reader role.

Reply:
Um, I don't understand your comment

Reply:
You pass unsanitized input to `system` (I realize right now we constrain the role to be one of `train`, `validate`, or `test` when the data reader is set up via prototext). I did notice that particular role name won't quite work; `; rm -rf --no-preserve-root /*` will do it instead.

`system` should basically never be used outside of quick debugging hacks.